File: 2024/day04_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    matrix = [row.strip() for row in content[::]]
    return sum(
        count(matrix, row, col)
        for row in range(len(matrix))
        for col in range(len(matrix[0]))
        if matrix[row][col] == "X"
    )


def count(matrix: list[str], row: int, col: int):
    return len(
        [
            check
            for check in [
                # Horizontal Forward
                (
                    col < len(matrix[0]) - 3
                    and matrix[row][col + 1] == "M"
                    and matrix[row][col + 2] == "A"
                    and matrix[row][col + 3] == "S"
                ),
                # Horizontal Backward
                (
                    col > 2
                    and matrix[row][col - 1] == "M"
                    and matrix[row][col - 2] == "A"
                    and matrix[row][col - 3] == "S"
                ),
                # Vertical Downward
                (
                    row < len(matrix) - 3
                    and matrix[row + 1][col] == "M"
                    and matrix[row + 2][col] == "A"
                    and matrix[row + 3][col] == "S"
                ),
                # Vertical Upward
                (
                    row > 2
                    and matrix[row - 1][col] == "M"
                    and matrix[row - 2][col] == "A"
                    and matrix[row - 3][col] == "S"
                ),
                # Diagonal TL-BR
                (
                    row < len(matrix) - 3
                    and col < len(matrix[0]) - 3
                    and matrix[row + 1][col + 1] == "M"
                    and matrix[row + 2][col + 2] == "A"
                    and matrix[row + 3][col + 3] == "S"
                ),
                # Diagonal BR-TL
                (
                    row > 2
                    and col > 2
                    and matrix[row - 1][col - 1] == "M"
                    and matrix[row - 2][col - 2] == "A"
                    and matrix[row - 3][col - 3] == "S"
                ),
                # Diagonal BL-TR
                (
                    row > 2
                    and col < len(matrix[0]) - 3
                    and matrix[row - 1][col + 1] == "M"
                    and matrix[row - 2][col + 2] == "A"
                    and matrix[row - 3][col + 3] == "S"
                ),
                # Diagonal TR-BL
                (
                    row < len(matrix) - 3
                    and col > 2
                    and matrix[row + 1][col - 1] == "M"
                    and matrix[row + 2][col - 2] == "A"
                    and matrix[row + 3][col - 3] == "S"
                ),
            ]
            if check
        ]
    )


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2024/day04-1.ts
type Coords = {
  row: number;
  col: number;
};

function solution(content: string) {
  const matrix = content.split("\n").filter((line) => line);
  const xes = matrix.reduce(
    (rowAcc, row, rowIndex) => [
      ...rowAcc,
      ...row
        .split("")
        .reduce(
          (colAcc, col, colIndex) =>
            col === "X"
              ? [...colAcc, { col: colIndex, row: rowIndex }]
              : colAcc,
          [] as Coords[]
        ),
    ],
    [] as Coords[]
  );
  return xes.reduce((acc, cur) => acc + count(matrix, cur), 0);
}

function count(matrix: string[], { row, col }: Coords) {
  return [
    // Horizontal forward
    col < matrix[0].length - 3 &&
      matrix[row][col + 1] === "M" &&
      matrix[row][col + 2] === "A" &&
      matrix[row][col + 3] === "S",
    // Horizontal backward
    col > 2 &&
      matrix[row][col - 1] === "M" &&
      matrix[row][col - 2] === "A" &&
      matrix[row][col - 3] === "S",
    // Vertical Downward
    row < matrix.length - 3 &&
      matrix[row + 1][col] === "M" &&
      matrix[row + 2][col] === "A" &&
      matrix[row + 3][col] === "S",
    // Vertical Upward
    row > 2 &&
      matrix[row - 1][col] === "M" &&
      matrix[row - 2][col] === "A" &&
      matrix[row - 3][col] === "S",
    // Diagonal TL-BR
    row < matrix.length - 3 &&
      col < matrix[0].length - 3 &&
      matrix[row + 1][col + 1] === "M" &&
      matrix[row + 2][col + 2] === "A" &&
      matrix[row + 3][col + 3] === "S",
    // Diagonal BR-TL
    row > 2 &&
      col > 2 &&
      matrix[row - 1][col - 1] === "M" &&
      matrix[row - 2][col - 2] === "A" &&
      matrix[row - 3][col - 3] === "S",
    // Diagonal BL-TR
    row > 2 &&
      col < matrix[0].length - 3 &&
      matrix[row - 1][col + 1] === "M" &&
      matrix[row - 2][col + 2] === "A" &&
      matrix[row - 3][col + 3] === "S",
    //Diagonal TR-BL
    row < matrix.length - 3 &&
      col > 2 &&
      matrix[row + 1][col - 1] === "M" &&
      matrix[row + 2][col - 2] === "A" &&
      matrix[row + 3][col - 3] === "S",
  ].filter((item) => item).length;
}

Deno.readTextFile("day04.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
